### PR TITLE
plan: remove aggregation grouped by unique key.

### PR DESCRIPTION
**Pull request by winoros**
Aggregations grouped by unique key can be substitute by projection.
PTAL @hanfei1991 @coocood @shenli @lamxTyler

### CODE REVIEW

File: plan/aggregation_pruning.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: 2017

File: plan/aggregation_pruning.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+type aggPrune struct {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: Pruner is better

File: plan/aggregation_pruning.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+				}
+				// GroupByCols has unique key. So this aggregation can be removed.
+				proj := &Projection{
+					Exprs:           make([]expression.Expression, 0, len(agg.AggFuncs)),
+					baseLogicalPlan: newBaseLogicalPlan(Proj, ap.allocator),
+				}
+				proj.self = proj
+				proj.initIDAndContext(ap.ctx)
+				for _, fun := range agg.AggFuncs {
+					proj.Exprs = append(proj.Exprs, fun.GetArgs()[0].Clone())
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: It's definitly wrong to do this.
1. For count, count(arg) should be transformed to if (arg is null, 0, 1).
2. For sum, we should consider the type casting: If arg is string, it should be cast(arg as double).

So add test to executor_test.go

File: plan/aggregation_pruning.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type aggPruner struct {
+	allocator *idAllocator
+	ctx       context.Context
+}
+
+func (ap *aggPruner) pruneAggregation(p LogicalPlan) error {
```
Comment (shenli) [inline plan/aggregation_pruning.go]: eliminateAggregation seems a better name.
Please add some comments for this function.

File: plan/aggregation_pruning.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+// eliminateAggregation will eliminate aggregation grouped by unique key.
+// e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
+// For count(expr), sum(expr), avg(expr), we may need to rewrite the expr. Details is shown below.
+func (ap *aggPruner) eliminateAggregation(p LogicalPlan) error {
+	for _, child := range p.GetChildren() {
+		if agg, ok := child.(*Aggregation); ok {
+			schemaByGroupby := expression.NewSchema(agg.groupByCols)
+			coveredByUniqueKey := false
+			for _, key := range agg.schema.Keys {
```
Comment (shenli) [inline plan/aggregation_pruning.go]: for table "create table t (c1 int, c2 int unique index);" and query "select count(c1) from t group by c2;" , c2 is a uniq key but it is not in agg.schema.

Comment (winoros) [inline plan/aggregation_pruning.go]: c2 is not in schema.Columns, but [c2] is in schema.Keys.
https://github.com/pingcap/tidb/blob/master/plan/build_key_info.go#L55 This part make sure that unique keys in group by will stay in the schema.Keys.

**Comment by winoros** [comment]
PTAL @hanfei1991 @shenli

File: plan/aggregation_pruning.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+				continue
+			}
+			// GroupByCols has unique key. So this aggregation can be removed.
+			proj := &Projection{
+				Exprs:           make([]expression.Expression, 0, len(agg.AggFuncs)),
+				baseLogicalPlan: newBaseLogicalPlan(Proj, ap.allocator),
+			}
+			proj.self = proj
+			proj.initIDAndContext(ap.ctx)
+			for _, fun := range agg.AggFuncs {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: this function is too long, abstract it as a function.

File: plan/logical_plan_test.go [hunk 7]
```diff
@@ -1459,23 +1459,23 @@ func (s *testPlanSuite) TestUniqueKeyInfo(c *C) {
 		{
 			sql: "select a, sum(e) from t group by a",
 			ans: map[string][][]string{
-				"TableScan_1":   {{"test.t.a"}},
+				"TableScan_1":   {{"test.t.f"}, {"test.t.g"}, {"test.t.f", "test.t.g"}, {"test.t.a"}},
```
Comment (hanfei1991) [inline plan/logical_plan_test.go]: Why change these tests?

**Comment by hanfei1991** [comment]
Add a test that output the plan string.

File: plan/logical_plan_test.go (hunk 7 above)
Comment (winoros) [inline plan/logical_plan_test.go]: The position of calling eliminateAggregation is chagned in optimizer.go. Now it's before PrunColumns. So key is more than before.

Comment (hanfei1991) [inline plan/logical_plan_test.go]: You can do pruning in this test.

File: plan/aggregation_pruning.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+				expr, err := ap.rewriteExpr(fun.GetArgs()[0].Clone(), fun.GetName())
+				if err != nil {
+					return errors.Trace(err)
+				}
+				proj.Exprs = append(proj.Exprs, expr)
+			}
+			proj.SetSchema(agg.schema.Clone())
+			InsertPlan(p, agg, proj)
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: using `SetChildren` and `SetParents` is more reasonable.

File: plan/aggregation_pruning.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+type aggPruner struct {
+	allocator *idAllocator
+	ctx       context.Context
+}
+
+// eliminateAggregation will eliminate aggregation grouped by unique key.
+// e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
+// For count(expr), sum(expr), avg(expr), we may need to rewrite the expr. Details is shown below.
+func (ap *aggPruner) eliminateAggregation(p LogicalPlan) error {
+	for _, child := range p.GetChildren() {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: It's not intuitive to process children. You can return a LogicalPlan instead and process p itself.

File: plan/aggregation_pruning.go [hunk 10]
```diff
@@ -29,7 +29,8 @@ type aggPruner struct {
 // eliminateAggregation will eliminate aggregation grouped by unique key.
 // e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
 // For count(expr), sum(expr), avg(expr), we may need to rewrite the expr. Details is shown below.
-func (ap *aggPruner) eliminateAggregation(p LogicalPlan) error {
+func (ap *aggPruner) eliminateAggregation(p LogicalPlan) (LogicalPlan, error) {
+	newChildren := make([]Plan, 0, len(p.GetChildren()))
 	for _, child := range p.GetChildren() {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: If you return LogicalPlan, you can call eliminateAggregation for every child. And only process `ap` here.

File: plan/aggregation_pruning.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	if agg, ok := p.(*Aggregation); ok {
+		schemaByGroupby := expression.NewSchema(agg.groupByCols)
+		coveredByUniqueKey := false
+		for _, key := range agg.schema.Keys {
+			if schemaByGroupby.GetColumnsIndices(key) != nil {
+				coveredByUniqueKey = true
+				break
+			}
+		}
+		if coveredByUniqueKey {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: `if !covered` then continue.

File: plan/aggregation_pruning.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+			RetType: types.NewFieldType(mysql.TypeLonglong),
+		}
+		newExpr, err = expression.NewFunction(ast.If, zero.RetType, isNullExpr, zero, one)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	case ast.AggFuncSum:
+		// Numeric type do nothing. Others will cast to decimal.
+		switch expr.GetType().Tp {
+		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeDouble,
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: Should we cast int to double ?

File: plan/aggregation_pruning.go (hunk 11 above)
Comment (winoros) [inline plan/aggregation_pruning.go]: But there is no loop so continue doesn't work. Change the outside if statement to switch then use break?

Comment (hanfei1991) [inline plan/aggregation_pruning.go]: never mind, that's ok

**Comment by shenli** [comment]
@winoros Any update? Please resolve the conflicts.

File: plan/aggregation_pruning.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+// rewriteExpr will rewrite the aggregate function to expression doesn't contain aggregate function.
+func (ap *aggPruner) rewriteExpr(expr expression.Expression, funcName string) (newExpr expression.Expression, err error) {
+	switch funcName {
+	case ast.AggFuncCount:
+		// If is count(expr), we will change it to if(isnull(expr), 0, 1).
+		isNullExpr, err := expression.NewFunction(ap.ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), expr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		zero := &expression.Constant{
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: you can use expression.Zero now

File: plan/aggregation_pruning.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+			Value:   types.NewIntDatum(0),
+			RetType: types.NewFieldType(mysql.TypeLonglong),
+		}
+		one := &expression.Constant{
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: ditto

File: plan/aggregation_pruning.go [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+			RetType: types.NewFieldType(mysql.TypeLonglong),
+		}
+		newExpr, err = expression.NewFunction(ap.ctx, ast.If, zero.RetType, isNullExpr, zero, one)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	case ast.AggFuncSum, ast.AggFuncAvg:
+		switch expr.GetType().Tp {
+		// Float-point number or fixed-point numnber will do nothing, Others will cast to decimal.
+		case mysql.TypeDouble, mysql.TypeFloat, mysql.TypeDecimal, mysql.TypeNewDecimal:
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: According to https://github.com/pingcap/tidb/blob/master/expression/util.go#L73 Should we cast int to decimal and others to double ?

File: plan/aggregation_pruning.go (hunk 13 above)
Comment (winoros) [inline plan/aggregation_pruning.go]: expression.Zero is TypeTiny, but count should use TypeLongLong as RetType.  I think this will make something wrong.

Comment (winoros) [inline plan/aggregation_pruning.go]: expression.Zero is TypeTiny, but count should use TypeLongLong as RetType. Will this change make something wrong?

File: plan/aggregation_pruning.go (hunk 15 above)
Comment (winoros) [inline plan/aggregation_pruning.go]: I just tried this but i found that currently CastFuncFactory doesn't support cast to double.
Should i add `cast to double ` case into CastFuncFactory?

File: plan/aggregation_pruning.go (hunk 13 above)
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: It doesn't matter

File: plan/aggregation_pruning.go (hunk 15 above)
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: yeah, you can add it

File: plan/aggregation_pruning.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+type aggPruner struct {
+	allocator *idAllocator
+	ctx       context.Context
+}
+
+// eliminateAggregation will eliminate aggregation grouped by unique key.
+// e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
+// For count(expr), sum(expr), avg(expr), we may need to rewrite the expr. Details is shown below.
```
Comment (zimulala) [inline plan/aggregation_pruning.go]: s/is/are

File: plan/aggregation_pruning.go [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+		if coveredByUniqueKey {
+			// GroupByCols has unique key. So this aggregation can be removed.
+			proj := &Projection{
+				Exprs:           make([]expression.Expression, 0, len(agg.AggFuncs)),
+				baseLogicalPlan: newBaseLogicalPlan(Proj, ap.allocator),
+			}
+			proj.self = proj
+			proj.initIDAndContext(ap.ctx)
+			for _, fun := range agg.AggFuncs {
+				expr, err := ap.rewriteExpr(fun.GetArgs()[0].Clone(), fun.GetName())
```
Comment (zimulala) [inline plan/aggregation_pruning.go]: Why only use the first argument？

File: plan/aggregation_pruning.go [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+		// If is count(expr), we will change it to if(isnull(expr), 0, 1).
+		isNullExpr, err := expression.NewFunction(ap.ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), expr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		newExpr, err = expression.NewFunction(ap.ctx, ast.If, types.NewFieldType(mysql.TypeLonglong), isNullExpr, expression.Zero, expression.One)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	// https://dev.mysql.com/doc/refman/5.7/en/group-by-functions.html
```
Comment (zimulala) [inline plan/aggregation_pruning.go]: Add  `See `.

File: plan/aggregation_pruning.go [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+	// The SUM() and AVG() functions return a DECIMAL value for exact-value arguments (integer or DECIMAL),
```
Comment (zimulala) [inline plan/aggregation_pruning.go]: s/DECIMAL/decimal
Unify `integer` and `decimal` formats. Update the similar lines.

Comment (winoros) [inline plan/aggregation_pruning.go]: Here `integer` is not one certain type, it includes `TINYINT`, `INT`, `INTEGER` ans so on, but `DECIMAL` is a certain type. So one is lowercase while another is uppercase.

Comment (zimulala) [inline plan/aggregation_pruning.go]: I got it.

File: plan/aggregation_pruning.go (hunk 17 above)
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: Count distinct may accept mutiple arguments, so you need to check it as If arg0 is not null and arg1 is not null and arg2 ....

File: plan/aggregation_pruning.go [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+		isNullExprs := make([]expression.Expression, 0, len(exprs))
+		for _, expr := range exprs {
+			isNullExpr, err := expression.NewFunction(ap.ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), expr.Clone())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			isNullExprs = append(isNullExprs, isNullExpr)
+		}
+		var innerExpr expression.Expression
+		if len(isNullExprs) > 1 {
```
Comment (hanfei1991) [inline plan/aggregation_pruning.go]: You can use `ComposeDNF`

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@zimulala PTAL

File: plan/aggregation_pruning.go [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+	retPlan.SetChildren(newChildren...)
+	return retPlan, nil
+}
+
+// rewriteExpr will rewrite the aggregate function to expression doesn't contain aggregate function.
+func (ap *aggPruner) rewriteExpr(exprs []expression.Expression, funcName string) (newExpr expression.Expression, err error) {
+	switch funcName {
+	case ast.AggFuncCount:
+		// If is count(expr), we will change it to if(isnull(expr), 0, 1).
+		// If is count(distinct x, y, z) we will change it to if(or(or(isnull(x), isnull(y)), isnull(z)), 0, 1).
```
Comment (zimulala) [inline plan/aggregation_pruning.go]: `or(isnull(x), isnull(y))` -> `isnull(x) or isnull(y)` ?
MySQL doesn't have the function of `or`.

**Comment by zimulala** [comment]
Rest LGTM

**Comment by zimulala** [comment]
Fix CI @winoros

**Pull request merged**